### PR TITLE
feat: implement OOO acknowledgement functionality

### LINKED ISSUE #2472: feat: implement OOO acknowledgement functionality
## Issue Description :
  - Implement functionality to acknowledge Out-of-Office (OOO) requests, allowing authorized users to acknowledge them.

## Current Behaviour
-  Users cannot create OOO requests or have them acknowledged. The OOO acknowledgement flow is blocked/missing, and related tests are currently skipped or commented.

## Expected Behaviour

-  Users can submit OOO requests with type OOO, from, until, and reason.
- Authorized users can acknowledge OOO requests:
- Approve or reject with optional comment.
- Validation prevents acknowledging non-OOO requests, already-approved, or already-rejected requests.
- System updates the request status, persists lastModifiedBy, and timestamps changes.

### PR DESCRIPTION
**Pull request by RishiChaubey31**
Date: 25-08-2025

Developer Name: Rishi Chaubey

---

## Issue Ticket Number

- closes #2472 

## Description

1. When a user creates OOO  request it has to be acknowledged by a SUPERUSER
2. Superuser can acknowledge or reject the OOO request.
3. The system tracks acknowledgements and updates the OOO status accordingly.
4. If all required acknowledgements are received, the OOO status is marked as acknowledged.

 

### Documentation Updated?

- [x] Yes
- [ ] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

https://github.com/user-attachments/assets/8d289c4e-6c82-4b27-9b0c-8e2ad779803d

Screenshot 2025-09-01 035209

Screenshot 2025-09-01 035340

</details>

## Test Coverage
- tests https://github.com/Real-Dev-Squad/website-backend/pull/2474
<details>

<summary>Screenshot 1</summary>

image

Screenshot 2025-09-03 011250

Screenshot 2025-09-03 011039

</details>

### What change is being made?

Implement the functionality to acknowledge Out-of-Office (OOO) requests by adding acknowledgment routes, controllers, services, validations, and necessary constants updates.

### Why are these changes being made?

This change introduces the capability to acknowledge OOO requests, allowing authorized superusers to approve or reject the requests while maintaining strict validation and role-based access control. The added functionality helps ensure requests are processed efficiently and securely, providing clear error handling and validation feedback across the system. Adding this feature enhances request management, aligning with current user roles and operational requirements, although it adds complexity that needs proper testing and review to handle potential edge cases.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: controllers/oooRequests.ts [hunk 1]
```diff
@@ -148,3 +153,45 @@
...
+
+      return res.status(200).json({
+        message: response.message,
+      });
+    }
+    catch(error){
+      logger.error(ERROR_WHILE_ACKNOWLEDGING_REQUEST, error);
+      next(error);
+      return res;
+  }
```
[the remark this answers is not in the document]
Comment (RishiChaubey31) [reply controllers/oooRequests.ts:196]: The comment is wrong — next(error) returns void but my function needs to return OooRequestResponse, so I’ll keep the original next(error); return res; pattern.

**Review by Suvidh-kaushik — CHANGES_REQUESTED** [review]

File: controllers/oooRequests.ts [hunk 2]
```diff
@@ -148,3 +152,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
     return res.boom.badImplementation(ERROR_WHILE_UPDATING_REQUEST);
   }
 };
+/**
+ * Acknowledges an Out-of-Office (OOO) request
+ * 
+ * @param {AcknowledgeOooRequest} req - The request object.
+ * @param {OooRequestResponse} res - The response object.
+ * @returns {Promise<OooRequestResponse>} Resolves with success or failure.
+ */
```
Comment (MayankBansal12) [inline controllers/oooRequests.ts:161]: can we make it more specific?

File: controllers/oooRequests.ts [hunk 3]
```diff
@@ -148,3 +152,45 @@ export const updateOooRequestController = async (req: UpdateRequest, res: Custom
...
+)
+  : Promise<OooRequestResponse> => {
+    try {
+      const dev = req.query.dev === "true";
+      if(!dev) return res.boom.notImplemented("Feature not implemented");
+
+      const isSuperuser = req.userData?.roles?.super_user;
+      if (!isSuperuser) {
+        return res.boom.forbidden(UNAUTHORIZED_TO_UPDATE_REQUEST);
+      }
```
Comment (MayankBansal12) [inline controllers/oooRequests.ts:175]: please use existing middleware for these checks and remove from here

File: middlewares/validators/requests.ts [hunk 4]
```diff
@@ -92,6 +92,10 @@ export const getRequestsMiddleware = async (req: OooRequestCreateRequest, res: O
       .string()
       .valid(REQUEST_STATE.APPROVED, REQUEST_STATE.PENDING, REQUEST_STATE.REJECTED)
       .optional(),
+    status: joi
+      .string()
+      .valid(REQUEST_STATE.APPROVED, REQUEST_STATE.PENDING, REQUEST_STATE.REJECTED)
+      .optional(),
     page: joi.number().integer().min(0).when("next", {
```
[the remark this answers is not in the document]
Comment (MayankBansal12) [reply middlewares/validators/requests.ts:99]: @RishiChaubey31 Please put a ToDo comment to cleanup state while removing dev feature flag

File: services/oooRequest.ts [hunk 5]
```diff
@@ -92,3 +99,98 @@ export const createOooRequest = async (
...
+                requestId,
+                status: userState.OOO,
+                state: statusState.UPCOMING,
+                from: requestData.from,
+                endsOn: requestData.until,
+                userId: requestData.requestedBy,
+                message: body.comment,
+                createdAt: Date.now()
+            });
+        }
```
[the remark this answers is not in the document]
Comment (RishiChaubey31) [reply services/oooRequest.ts:184]: When a user goes OOO, their current status might be ACTIVE/IDLE, but they need a future status of OOO hence we require addFutureStatus and the createFutureStatus keeping log and tracking of how many OOO requests does user take  in any period of time

File: constants/requests.ts [hunk 6]
```diff
@@ -26,6 +26,8 @@ export const REQUEST_LOG_TYPE = {
   REQUEST_CANCELLED: "REQUEST_CANCELLED",
   REQUEST_UPDATED: "REQUEST_UPDATED",
   PENDING_REQUEST_FOUND: "PENDING_REQUEST_FOUND",
+  REQUEST_ALREADY_APPROVED: "REQUEST_ALREADY_APPROVED",
+  REQUEST_ALREADY_REJECTED: "REQUEST_ALREADY_REJECTED",
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply constants/requests.ts:30]: but the message content is same right, then why to have redundant constant, constants are created so that we can repeatedly use them in different places wherever required

File: controllers/oooRequests.ts [hunk 7]
```diff
@@ -78,7 +83,7 @@ export const createOooRequestController = async (
         return res.boom.conflict(REQUEST_ALREADY_PENDING);
     }
 
-    await createOooRequest(requestBody, username, userId);
+    await createOooRequest(requestBody, userId);
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply controllers/oooRequests.ts:86]: korbit-ai , check again some new code got pushed for create service

File: models/requests.ts [hunk 8]
```diff
@@ -84,8 +84,16 @@ export const getRequests = async (query: any) => {
     if (id) {
       const requestDoc = await requestModel.doc(id).get();
       if (!requestDoc.exists) {
+
```
Comment (Suvidh-kaushik) [inline models/requests.ts:87]: what is this for?

File: routes/requests.ts [hunk 9]
```diff
@@ -22,6 +23,6 @@ import { verifyDiscordBot } from "../middlewares/authorizeBot";
 router.get("/", getRequestsMiddleware, getRequestsController);
 router.post("/", skipAuthenticateForOnboardingExtensionRequest(authenticate, verifyDiscordBot), createRequestsMiddleware, createRequestController);
 router.put("/:id",authenticate, authorizeRoles([SUPERUSER]), updateRequestsMiddleware, updateRequestController);
-router.patch("/:id", authenticate, updateRequestValidator, updateRequestBeforeAcknowledgedController);
+router.patch("/:id", authenticate, conditionalOooChecks, updateRequestValidator, updateRequestBeforeAcknowledgedController);
```
Comment (Suvidh-kaushik) [inline routes/requests.ts:26]: please give a better name instead of conditionalOooChecks

**Review by Suvidh-kaushik — CHANGES_REQUESTED** [review]

File: models/requests.ts (hunk 8 above)
Comment (RishiChaubey31) [reply models/requests.ts:87]: getRequestById was a separate function earlier for ooo, but since it got removed, its behavior was merged into the unified getRequests function using conditional logic.

File: routes/requests.ts (hunk 9 above)
Comment (RishiChaubey31) [reply routes/requests.ts:26]: Could you please suggest a better name that you think would be more appropriate for this file?

**Comment by AnujChhikara** [comment]
@RishiChaubey31 please attach the working proof that user status is getting changed after approving the request

File: middlewares/validators/requests.ts [hunk 10]
```diff
@@ -88,7 +88,11 @@ export const getRequestsMiddleware = async (req: OooRequestCreateRequest, res: O
       .valid(REQUEST_TYPE.OOO, REQUEST_TYPE.EXTENSION, REQUEST_TYPE.TASK, REQUEST_TYPE.ALL, REQUEST_TYPE.ONBOARDING)
       .optional(),
     requestedBy: joi.string().insensitive().optional(),
-    state: joi
+    state: joi // TODO: Remove this validator once feature is tested and ready to be used
```
Comment (AnujChhikara) [inline middlewares/validators/requests.ts:91]: - why can't we remove this in PR as we have put this route behind the dev flag and in new structure state field is not required

File: models/requests.ts [hunk 11]
```diff
@@ -84,8 +84,16 @@ export const getRequests = async (query: any) => {
     if (id) {
       const requestDoc = await requestModel.doc(id).get();
       if (!requestDoc.exists) {
+        
+        if (type === 'OOO') {
```
Comment (AnujChhikara) [inline models/requests.ts:88]: - why this check as we are fetching request by id and I don't think this check needed as if no request found with that particular id just throw the error not found

File: services/oooRequest.ts [hunk 12]
```diff
@@ -51,22 +60,20 @@ export const validateUserStatus = async (
  * Create an OOO request for a user.
  * 
  * @param {OooStatusRequestBody} body - The request body containing OOO details.
- * @param {string} username - The username of the person creating the request.
- * @param {string} userId - The unique identifier of the user.
+ * @param {string} requestedBy - The userid of the person creating the request.
```
Comment (AnujChhikara) [inline services/oooRequest.ts:63]: nit: replace userid with userId

File: services/oooRequest.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnujChhikara) [reply services/oooRequest.ts:184]: - so where we are setting or changing the user status?

File: models/requests.ts [hunk 13]
```diff
@@ -84,8 +84,16 @@ export const getRequests = async (query: any) => {
...
+          throw new NotFound(REQUEST_DOES_NOT_EXIST);
+        }
         return null;
       }
+      
+      if (type === 'OOO') {
+        return requestDoc.data();
+      }
```
Comment (MayankBansal12) [inline models/requests.ts:96]: - is there a need for OOO type check here?
since this is already included in select query?

File: models/requests.ts (hunk 11 above)
Comment (MayankBansal12) [reply models/requests.ts:88]: +1

File: services/oooRequest.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (RishiChaubey31) [reply services/oooRequest.ts:184]: ## Functions

### `addFutureStatus`
- Updates the user's status document in the **`usersStatus`** collection.  
- Adds the **OOO** status to the `futureStatus` field.  

### `createUserFutureStatus`
- Creates a separate record in the **`userFutureStatus`** collection.  
- Used for tracking future status changes.  


After that the system has an automated process that runs via a cron job (accessible via the /userStatus/update endpoint) that calls updateAllUserStatus

**Comment by RishiChaubey31** [comment]
> @RishiChaubey31 please attach the working proof that user status is getting changed after approving the request

I have added the proofs in screenshots

File: controllers/oooRequests.ts [hunk 14]
```diff
@@ -41,11 +44,13 @@ export const createOooRequestController = async (
...
-  const { id: userId, username } = req.userData;
-  const isUserPartOfDiscord = req.userData.roles.in_discord;
+  const { id: userId, roles } = req.userData;
   const dev = req.query.dev === "true";
+  const isUserPartOfDiscord = roles.in_discord;
 
-  if (!dev) return res.boom.notImplemented("Feature not implemented");
+  if (!dev) {
+    return res.boom.notImplemented("Feature not implemented");
+  }
```
[the remark this answers is not in the document]
Comment (RishiChaubey31) [reply controllers/oooRequests.ts:53]: Using the dev middleware here I think would make no sense as it would increase the complexity of the code , its just a simple check

**Review by MayankBansal12 — DISMISSED** [review]

**Review by Suvidh-kaushik — APPROVED** [review]

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: middlewares/validators/oooRequests.ts [hunk 15]
```diff
@@ -38,3 +38,58 @@ export const createOooStatusRequestValidator = async (
...
+  req: AcknowledgeOooRequest,
+  res: OooRequestResponse,
+  next: NextFunction
+): Promise<void> => {
+  try {
+    await acknowledgeOooRequestSchema.validateAsync(req.body, { abortEarly: false });
+    await paramsSchema.validateAsync(req.params, { abortEarly: false });
+    next();
+  } catch (error) {
+    const errorMessages = error.details.map((detail) => detail.message);
```
Comment (iamitprakash) [inline middlewares/validators/oooRequests.ts:91]: error.details may not exist

reason - 
Joi.validateAsync throws a ValidationError (which has .details), but if any other error occurs, error.details will be undefined.

Calling error.details.map(...) without checking will cause a runtime error.

File: middlewares/validators/oooRequests.ts [hunk 16]
```diff
@@ -38,3 +38,58 @@ export const createOooStatusRequestValidator = async (
...
+
+/**
+ * Middleware to validate the acknowledge Out-Of-Office (OOO) request payload.
+ * 
+ * @param {AcknowledgeOooRequest} req - The request object containing the body to be validated.
+ * @param {OooRequestResponse} res - The response object used to send error responses if validation fails.
+ * @param {NextFunction} next - The next middleware function to call if validation succeeds.
+ * @returns {Promise<void>} Resolves or sends errors.
+ */
+export const acknowledgeOooRequestValidator = async (
```
Comment (iamitprakash) [inline middlewares/validators/oooRequests.ts:81]: this will break at runtime is error.details.map(...) without checking if error is a Joi error.

File: middlewares/validators/oooRequests.ts [hunk 17]
```diff
@@ -38,3 +38,58 @@ export const createOooStatusRequestValidator = async (
...
+  req: AcknowledgeOooRequest,
+  res: OooRequestResponse,
+  next: NextFunction
```
Comment (iamitprakash) [inline middlewares/validators/oooRequests.ts:84]: you typed req: AcknowledgeOooRequest and res: OooRequestResponse.

Unless those extend Request and Response from Express properly, middleware may not type-check correctly.

Especially res.boom—that implies express-boom is being used, so OooRequestResponse must extend Response & { boom: ... }.

File: middlewares/oooRoleCheckMiddleware.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * @param {CustomRequest} req - The request object
+ * @param {CustomResponse} res - The response object  
+ * @param {NextFunction} next - The next middleware function
+ */
+export const oooRoleCheckMiddleware = (req: CustomRequest, res: CustomResponse, next: NextFunction) => {
+  const requestType = req.body?.type;
+
+
+  if (requestType === REQUEST_TYPE.OOO) {
+     //TODO: Remove this middleware once the OOO feature is tested and ready to be used
```
Comment (iamitprakash) [inline middlewares/oooRoleCheckMiddleware.ts:21]: replace with below

```
import { NextFunction } from "express";
import { CustomRequest, CustomResponse } from "../types/global";
import { REQUEST_TYPE } from "../constants/requests";
import { devFlagMiddleware } from "./devFlag";
import authorizeRoles from "./authorizeRoles";

const { SUPERUSER } = require("../constants/roles");

export const oooRoleCheckMiddleware = (
  req: CustomRequest,
  res: CustomResponse,
  next: NextFunction
) => {
  const requestType = req.body?.type;

  if (requestType === REQUEST_TYPE.OOO) {
    // TODO: Remove this middleware once the OOO feature is tested and ready 
    return devFlagMiddleware(req, res, (err?: any) => {
      if (err) return next(err);

      try {
        return authorizeRoles([SUPERUSER])(req, res, next);
      } catch (authErr) {
        return next(authErr);
      }
    });
  }

  // no check will by-pass this
  return next();
};

```

File: services/oooRequest.ts [hunk 19]
```diff
@@ -88,7 +98,103 @@ export const createOooRequest = async (
 
         return request;
     } catch (error) {
-        logger.error("Error while creating OOO request", error);
+        logger.error(ERROR_WHILE_CREATING_REQUEST, error);
+        throw error;
+    }
+}
+
+/**
```
Comment (iamitprakash) [inline services/oooRequest.ts:106]: replace with below

```
export const validateOooAcknowledgeRequest = (
  requestType: string,
  requestStatus: string,
) => {
  if (requestType !== REQUEST_TYPE.OOO && requestType !== REQUEST_TYPE.OOO_ACKNOWLEDGEMENT) {
    logger.error(`Invalid request type: ${requestType}`);
    throw new BadRequest(INVALID_REQUEST_TYPE);
  }

  if (requestStatus === REQUEST_STATE.APPROVED) {
    logger.error(`Request already approved`);
    throw new BadRequest(REQUEST_ALREADY_APPROVED);
  }

  if (requestStatus === REQUEST_STATE.REJECTED) {
    logger.error(`Request already rejected`);
    throw new BadRequest(REQUEST_ALREADY_REJECTED);
  }
};

```

File: services/oooRequest.ts [hunk 20]
```diff
@@ -88,7 +98,103 @@ export const createOooRequest = async (
...
+ * Acknowledges the OOO request.
+ * 
+ * @param {string} requestId - The id of the request to be acknowledged.
+ * @param {AcknowledgeOooRequestBody} body - The Acknowledgement request body.
+ * @param {string} superUserId - The id of the super user.
+ * @returns {Promise<object>} The acknowledged OOO request.
+ * @throws {Error} Throws an error if an issue occurs during acknowledgement.
+ */
+
+export const acknowledgeOooRequest = async (
```
Comment (iamitprakash) [inline services/oooRequest.ts:144]: unsafe cast of requestData: if getRequests returns an oldOooStatusRequest, destructuring it as OooStatusRequest will blow up.

```
export const acknowledgeOooRequest = async (
  requestId: string,
  body: AcknowledgeOooRequestBody,
  superUserId: string,
) => {
  try {
    const requestData = await getRequests({ id: requestId }) as OooStatusRequest | oldOooStatusRequest;
    if (!requestData) {
      throw new NotFound("Request not found");
    }

    // Check type safely instead of force-cast  :P
    const { type, status, from, until, requestedBy } = requestData;

    await validateOooAcknowledgeRequest(type, status);

    const requestResult = await updateRequest(requestId, body, superUserId, REQUEST_TYPE.OOO);
    if (requestResult.error) {
      throw new BadRequest(requestResult.error);
    }

    const [acknowledgeLogType, returnMessage] =
      requestResult.status === REQUEST_STATE.APPROVED
        ? [REQUEST_LOG_TYPE.REQUEST_APPROVED, REQUEST_APPROVED_SUCCESSFULLY]
        : [REQUEST_LOG_TYPE.REQUEST_REJECTED, REQUEST_REJECTED_SUCCESSFULLY];

    await addLog(
      acknowledgeLogType,
      {
        requestId,
        action: LOG_ACTION.UPDATE,
        userId: superUserId,
      },
      requestResult,
    );

    if (requestResult.status === REQUEST_STATE.APPROVED) {
      await addFutureStatus({
        requestId,
        state: REQUEST_TYPE.OOO,
        from,
        endsOn: until,
        userId: requestedBy,
        message: body.comment ?? "",
      });

      await createUserFutureStatus({
        requestId,
        status: userState.OOO,
        state: statusState.UPCOMING,
        from,
        endsOn: until,
        userId: requestedBy,
        message: body.comment ?? "",
        createdAt: Date.now(),
      });
    }

    return {
      message: returnMessage,
      request: requestResult, // optional: return full request for clarity [If needed]
    };
  } catch (error) {
    logger.error(ERROR_WHILE_ACKNOWLEDGING_REQUEST, { error });
    throw error;
  }
};

```

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: services/oooRequest.ts (hunk 20 above)
Comment (RishiChaubey31) [reply services/oooRequest.ts:145]: Applied the suggestion

File: services/oooRequest.ts (hunk 19 above)
Comment (RishiChaubey31) [reply services/oooRequest.ts:106]: Applied the suggestion

File: middlewares/oooRoleCheckMiddleware.ts (hunk 18 above)
Comment (RishiChaubey31) [reply middlewares/oooRoleCheckMiddleware.ts]: Applied the suggestion

File: middlewares/validators/oooRequests.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (RishiChaubey31) [reply middlewares/validators/oooRequests.ts:81]: updated it!

File: middlewares/validators/oooRequests.ts [hunk 21]
```diff
@@ -38,3 +38,67 @@ export const createOooStatusRequestValidator = async (
...
+
+    if (isJoiValidationError(error)) {
+      const errorMessages = error.details.map((detail) => detail.message);
+      logger.error(`${ERROR_WHILE_ACKNOWLEDGING_REQUEST}: ${errorMessages}`);
+      return res.boom.badRequest(errorMessages);
+    }
+    const genericMessage = typeof (error as any)?.message === "string" ? (error as any).message : String(error);
+    logger.error(`${ERROR_WHILE_ACKNOWLEDGING_REQUEST}: ${genericMessage}`);
+    return res.boom.badRequest([ERROR_WHILE_ACKNOWLEDGING_REQUEST]);
+  }
```
Comment (MayankBansal12) [inline middlewares/validators/oooRequests.ts:103]: can we simplify this? the current code looks confusing with too many type cast.

File: utils/requests.ts [hunk 22]
```diff
@@ -69,27 +69,20 @@ export const newOOOSchema = (request: oldOooStatusRequest) => ({
...
+    if (dev) {
+      return 'status' in request 
+        ? oldOOOSchema(request as OooStatusRequest)
+        : request;
+    }
+
+    return 'state' in request 
+      ? newOOOSchema(request as oldOooStatusRequest)
+      : request;
+  });
```
Comment (MayankBansal12) [inline utils/requests.ts:87]: can we refactor this more
eg:
```
if(dev && request.status){
   return oldOOOSchema(request as OooStatusRequest)
}

if(request.state){
  return newOOOSchema(request as oldOooStatusRequest)
}

return request;

```

File: services/oooRequest.ts [hunk 23]
```diff
@@ -88,7 +98,114 @@ export const createOooRequest = async (
...
+    body: AcknowledgeOooRequestBody,
+    superUserId: string,
+  ) => {
+    try {
+      const requestData = await getRequests({ id: requestId }) as OooStatusRequest | oldOooStatusRequest;
+      if (!requestData) {
+        throw new NotFound("Request not found");
+      }
+  
+      const { type, status, from, until, requestedBy } = requestData as OooStatusRequest;
```
[the remark this answers is not in the document]
Comment (MayankBansal12) [reply services/oooRequest.ts:156]: @RishiChaubey31 Kindly reply to this comment or address the suggestion

File: services/oooRequest.ts [hunk 24]
```diff
@@ -88,7 +98,114 @@ export const createOooRequest = async (
...
+        await createUserFutureStatus({
+          requestId,
+          status: userState.OOO,
+          state: statusState.UPCOMING,
+          from,
+          endsOn: until,
+          userId: requestedBy,
+          message: body.comment ?? "",
+          createdAt: Date.now(),
+        });
```
Comment (MayankBansal12) [inline services/oooRequest.ts:199]: addFutureStatus, createUserFutureStatus both seems similar,  what's the diff?

Also, in createUserFutureStatus, why are we passing state UPCOMING?

File: services/oooRequest.ts [hunk 25]
```diff
@@ -88,7 +98,114 @@ export const createOooRequest = async (
...
+ * @throws {Error} Throws an error if an issue occurs during acknowledgement.
+ */
+
+export const acknowledgeOooRequest = async (
+    requestId: string,
+    body: AcknowledgeOooRequestBody,
+    superUserId: string,
+  ) => {
+    try {
+      const requestData = await getRequests({ id: requestId }) as OooStatusRequest | oldOooStatusRequest;
```
Comment (MayankBansal12) [inline services/oooRequest.ts:151]: why are we using type cast here? If types are properly defined for getRequests, then shouldn't this be clear what type `requestData` will be?

File: services/oooRequest.ts (hunk 23 above)
Comment (MayankBansal12) [inline services/oooRequest.ts:156]: this can be handled in a better way.
instead of forcefully picking up status, check if `request.status` is present or `request.state` is and use that as status value

File: middlewares/validators/requests.ts [hunk 26]
```diff
@@ -142,6 +146,9 @@ export const updateRequestValidator = async (
             req, 
             res as OnboardingExtensionResponse, next);
           break;
+      case REQUEST_TYPE.OOO:
+          await acknowledgeOooRequestValidator(req, res as OooRequestResponse, next);
```
Comment (MayankBansal12) [inline middlewares/validators/requests.ts:150]: shouldn't we use `req as AcknowledgeOooRequest` here?

File: services/oooRequest.ts (hunk 24 above)
Comment (RishiChaubey31) [reply services/oooRequest.ts:199]: ### `addFutureStatus`
- Updates the user's status document in the **`usersStatus`** collection.  
- Adds the **OOO** status to the `futureStatus` field.  

### `createUserFutureStatus`
- Creates a separate record in the **`userFutureStatus`** collection.  
- Used for tracking future status changes.  


After that the system has an automated process that runs via a cron job (accessible via the /userStatus/update endpoint) that calls updateAllUserStatus

---

we are passing UPCOMING here because it's the state where the request has been approved but the OOO period hasn't started yet

File: services/oooRequest.ts (hunk 25 above)
Comment (RishiChaubey31) [reply services/oooRequest.ts:151]: Since the current (old) getRequests function returns any, we’re type casting here.

Comment (MayankBansal12) [reply services/oooRequest.ts:151]: can we define type in getRequests, that way we won't have to use typecast here. unless there's a specific reason to use any there?
